Azdevify eshopOnWeb

zedy-wj opened the pull request:
This pr is all changes to convert eShopOnWeb application into an Azd template.

You can currently deploy the eShopOnWeb application using `azd` by following the steps below:

1. azd up -t https://github.com/zedy-wj/eShopOnWeb.git -b Azdevify
2. Wait a moment for the resource deployment to complete, click the web endpoint and you will see:
![image](https://user-images.githubusercontent.com/81678720/197704328-df571202-68c0-430a-8423-5a579dbf890d.png)

@jongio, @rajeshkamal5050 for notification.

zedy-wj committed:
Azdevify eshopOnWeb

jongio commented on azure.yaml:
```diff
@@ -0,0 +1,14 @@
+        project: ./src/Web
+        language: csharp
+        host: appservice
+infra:
+    provider: ""
+    path: ""
+    module: ""
+pipeline:
+    provider: ""
```
These shouldn't be in the file. they are optional

jongio commented on infra/app/dbCatalog.bicep:
```diff
@@ -0,0 +1,25 @@
+param environmentName string
```
rename to catalog-db.bicep

jongio commented on infra/app/dbIdentity.bicep:
```diff
@@ -0,0 +1,25 @@
+param environmentName string
```
rename to identity-db.bicep

jongio commented on infra/app/web.bicep:
```diff
@@ -0,0 +1,18 @@
+
+param serviceName string = 'web'
+
+module web '../core/host/appservice-dotnet.bicep' = {
```
Make sure you have all the latest /bicep/core from azure-dev. It was recently updated.

jongio commented on infra/app/web.bicep:
```diff
@@ -0,0 +1,18 @@
+param serviceName string = 'web'
+
+module web '../core/host/appservice-dotnet.bicep' = {
+  name: '${serviceName}-appservice-dotnet-module'
```
Also, did you have to make any changes to bicep/core for this?

zedy-wj replied:
We created templates(`sqlserver-catalog.bicep` and `sqlserver-identity.bicep`) in sqlserver for `catalog` and `identity` separately, we didn't need to make any changes to bicep/core other than that.

jongio replied:
I removed these: https://github.com/zedy-wj/eShopOnWeb/pull/2

jongio commented:
To make this 100% compat with azd up then we need:

https://github.com/Azure/azure-dev/issues/958
https://github.com/Azure/azure-dev/issues/534

@zedy-wj - Please update the dotnet script above to not do the cd at the beginning, but use the full paths from the root of the repo.  Also, create a script that will do those, one in psh and one in .sh.

Add them to a /scripts folder.

Thanks!

jongio commented on .gitignore:
```diff
@@ -257,3 +257,5 @@ pub/
 
 #Ignore marker-file used to know which docker files we have.
 .eshopdocker_*
+
```
Please add the .devcontainer files as well.

zedy-wj committed:
Azdev-ify changes for latest /bicep/core

zedy-wj committed:
Azdev-ify changes for latest /bicep/core

zedy-wj committed:
Remove some invalid changes

zedy-wj committed:
add scripts folder and rename sqlserver

zedy-wj commented:
@jongio - We have updated according to your comments and moved some manual steps into the `scripts `folder, but replacing the database connectionString in the `appsettings.json` file still needs to be done manually.

jongio commented on .gitignore:
```diff
@@ -257,3 +257,6 @@ pub/
 
 #Ignore marker-file used to know which docker files we have.
 .eshopdocker_*
+.devcontainer
```
Please open this in codespaces.

1. It fails to load
2. Can you add azd to it?

Thanks!

zedy-wj replied:
We upgraded the `dotnet 5.0` version of the image in the dockerfile to `6.0` to solve the error when installing `dotnet-ef`. In addition, we added the steps to install `azd` in the dockerfile, and added azure-dev extention in `.devcontainer`.

jongio commented on infra/core/database/sqlserver/sqlserver-catalog.bicep:
```diff
@@ -0,0 +1,129 @@
+param name string
```
You don't need to create these files in core.  You can use the sqlserver.bicep file and pass the connectionStringKey to it.

jongio replied:
I resolved this in https://github.com/zedy-wj/eShopOnWeb/pull/2

zedy-wj replied:
Thanks! We have merged #2 into this pr.

jongio committed:
Update bicep add azure.yaml scripts

jongio commented on .gitignore:
```diff
@@ -257,3 +257,6 @@ pub/
 
 #Ignore marker-file used to know which docker files we have.
```
Please update the application to pull the connection strings from keyvault instead of appsettings.  That will eliminate the need to manually update the appsettings.json file after provision.

jongio commented:
Please install the azd cli from this pr to get the middleware hooks that are now in azure.yaml
https://github.com/Azure/azure-dev/pull/1256

zedy-wj committed:
add azd in codespace

zedy-wj committed:
upgrade dotnet core 5.0 to 6.0 in order to install dotnet-ef

zedy-wj committed:
install azd in codespace

zedy-wj committed:
install azd in codespace

zedy-wj committed:
Remove install azd in codespace, check some errors

zedy-wj committed:
Install azd in codespace, check some errors

zedy-wj committed:
Merge pull request #2 from jongio/Azdevify

Update bicep add azure.yaml scripts

zedy-wj committed:
eliminate manually update the appsettings.json file

zedy-wj commented:
@jongio  - We have done all the modifications according to your suggestion, no manual steps are needed now, we also tested on `codespace` and it still works fine.

The `dotnet ef database update` operation seems to be redundant, we don't need to run the `.ps1` file anymore, I will remove it and delete the `scripts` operation in `azure.yaml`, thanks for your help!

zedy-wj committed:
remove some invalid changes

jongio commented on .devcontainer/devcontainer.json:
```diff
@@ -8,10 +8,28 @@
         "NODE_VERSION": "lts/*",
-        "INSTALL_AZURE_CLI": "false"
+        "INSTALL_AZURE_CLI": "false",
+		"INSTALL_AZURE_DEVELOPER_CLI": "true"
```
Fix indent.

jongio commented on .devcontainer/devcontainer.json:
```diff
@@ -23,6 +41,7 @@
 
 	// Add the IDs of extensions you want installed when the container is created.
 	"extensions": [
+		"ms-azuretools.azure-dev",
```
Fix indent

jongio commented on infra/core/host/staticwebapp.bicep:
```diff
@@ -0,0 +1,21 @@
+param name string
```
Only include bicep files from core that are needed for this app.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,124 @@
+    runtimeVersion: '6.0'
+    tags: union(tags, { 'azd-service-name': 'web' })
+    appSettings: {
+      CATALOG_CONNECTION_STRING_VALUE: '${catalogDb.outputs.connectionString}; Password=${appUserPassword}'
+      IDENTITY_CONNECTION_STRING_VALUE: '${identityDb.outputs.connectionString}; Password=${appUserPassword}'
```
Please store the values in KV and then update the app to pull the values from KV.

jongio commented on scripts/setup-database.ps1:
```diff
@@ -0,0 +1,9 @@
+# Default to current directory
```
I don't follow your comment about not needing this file anymore.  Can you explain why it isn't needed?

zedy-wj committed:
fix indent

zedy-wj committed:
Fix indent

zedy-wj committed:
Remove invalid changes

zedy-wj committed:
Pull the connection string from key vault

zedy-wj committed:
fix some bug

zedy-wj commented:
@jongio - Currently we store `sqlConnectionString` in KV and retrieve it in app. At first, we tried to use `DefaultAzureCredential` for authentication, but `azd deploy` failed, the error message was: `stored credential not found`( we ensured that `tenantid`, `clientid` and `clientsecret` are configured locally and executed `az login`), so we switched to `ClientSecretCredential`authentication, it seems to work well, do you think this is suitable?

Besides, on why setting up the database file is no longer necessary: ​​we did a thorough test and no matter how they get the `SqlConnectionString` now from kv or previously from appsetting, they no longer need this step( dotnet ef database update ). At the same time, we also used the `dotnet run` command to start the app locally and found that the database configuration steps have been included.

The reason for the error here may be that we did not consider `azd deploy`, and it usually takes 3-5 minutes for the page to take effect. After adding this step (dotnet ef database update), the page that was not effective before is now effective, which led us to mistakenly think that this step is needed, but in fact, we did not give enough time to take effect before we checked it. Sorry for the confusion.

jongio commented on infra/main.bicep:
```diff
@@ -120,5 +121,14 @@ module appServicePlan './core/host/appserviceplan.bicep' = {
 }
 
+// Data outputs
+output AZURE_SQL_CATALOG_CONNECTION_STRING string = catalogDb.outputs.connectionStringKey
```
output AZURE_SQL_CATALOG_CONNECTION_STRING_KEY
output AZURE_SQL_IDENTITY_CONNECTION_STRING_KEY


since it isn't the cstring it is the key

jongio commented on src/Infrastructure/Dependencies.cs:
```diff
@@ -3,6 +3,9 @@
 using Microsoft.eShopWeb.Infrastructure.Identity;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
+using Azure.Security.KeyVault.Secrets;
```
Can you sort these usings?

jongio commented on src/Infrastructure/Dependencies.cs:
```diff
@@ -30,11 +39,24 @@ public static void ConfigureServices(IConfiguration configuration, IServiceColle
+            return "";
+        }
+
+        var secretClient = new SecretClient(new Uri(keyVaultUri), new ClientSecretCredential("<tenant_id>","<client_id>","<client_secret>"));
+        KeyVaultSecret secret = secretClient.GetSecret(connectionStringKey);
```
This should use DAC and also add the new AzureDeveloperCLICredential that can be found in azure-dev templates.  Look for the C# templates code on that.

jongio commented on src/Infrastructure/Dependencies.cs:
```diff
@@ -30,11 +39,24 @@ public static void ConfigureServices(IConfiguration configuration, IServiceColle
+
+        var secretClient = new SecretClient(new Uri(keyVaultUri), new ClientSecretCredential("<tenant_id>","<client_id>","<client_secret>"));
+        KeyVaultSecret secret = secretClient.GetSecret(connectionStringKey);
+        string secretValue = secret.Value;
+        return secretValue;
```
I don't think we need a var here, you could `return secret.Value`

zedy-wj committed:
Remove some invalid changes

zedy-wj committed:
Fix format and update the method of pulling connection string from kv with dac

zedy-wj commented:
@jongio - We have updated according to your comments, please review it, thanks!

jongio commented on src/Infrastructure/Dependencies.cs:
```diff
@@ -39,24 +31,11 @@ public static void ConfigureServices(IConfiguration configuration, IServiceColle
             services.AddDbContext<AppIdentityDbContext>(options =>
-                options.UseSqlServer(identityConnectionStringValue));
+                options.UseSqlServer(configuration.GetConnectionString("IdentityConnection")));
         }
```
Where does the app get the connection string from the key vault secret?

jongio commented on src/Web/Program.cs:
```diff
@@ -21,7 +23,19 @@
+builder.Configuration.AddAzureKeyVault(new Uri(builder.Configuration["AZURE_KEY_VAULT_ENDPOINT"]), credential);
+builder.Services.AddDbContext<CatalogContext>(c =>
+{
+    var connectionString = builder.Configuration[builder.Configuration["AZURE_SQL_CATALOG_CONNECTION_STRING_KEY"]];
```
Do we need 2 builder.Configuration statements?

How do we go from getting the key to getting the connection string from key vault?

zedy-wj replied:
The internal builder.Configuration will get the connection string key, and the `AddAzureKeyVault()` method will adds an `Microsoft.Extentions.Configuration.IConfigurationProvider` that reads configuration values from the Azure KeyVault. Therefore, the external builder.Configuration can get the corresponding secret value according to the key. For this implementation, we refer to `Azure-Samples/todo-csharp-sql` template, do you think this is appropriate? Our previous implementation was to add a method similar to `getSqlConnStr` to get the secret value from keyVault.

jongio replied:
Okay, I see what it is doing now.  Thanks,

jongio commented:
Can you please update this: https://github.com/zedy-wj/eShopOnWeb/pull/1#issue-1421933514 to reflect the current steps needed to get the app fully running with azd?  It is just `azd up` now?

zedy-wj commented:
@jongio - We have updated https://github.com/zedy-wj/eShopOnWeb/pull/1#issue-1421933514. Based on this [issue](https://github.com/Azure/azure-dev-pr/issues/1463#issue-1403543615), we have only tested `azd init` and `azd up` commands at the moment, what other azd commands do you think we need to test?

jongio commented:
Heads up that this also works:

`azd up -t https://github.com/zedy-wj/eShopOnWeb.git -b Azdevify`


You don't need to clone

zedy-wj commented:
@jongio - Thanks for the reminder, updated.

v-xuto commented:
@jongio We have updated the description and code based on your suggestion. Please re-review this PR.

jongio commented on scripts/setup-database.ps1:
```diff
@@ -0,0 +1,9 @@
+# Default to current directory
```
Do we still need the ps1 and sh filesin the repo?

zedy-wj replied:
We no longer need it, removed.

jongio commented:
It works for me!  Thanks a lot.

can you please open a PR against the upstream repo? https://github.com/dotnet-architecture/eShopOnWeb

Once we have that open, then I can ask the owners to review.

zedy-wj committed:
Remove invalid files

zedy-wj commented:
@jongio We opened a new pr: https://github.com/dotnet-architecture/eShopOnWeb/pull/844.
